File: obc_controller/ui/main_window.py
"""Main application window — 3-column space-themed layout."""

from __future__ import annotations

from pathlib import Path

from PySide6.QtCore import Qt, Slot
from PySide6.QtGui import QPixmap
from PySide6.QtWidgets import (
    QDialog,
    QFrame,
    QHBoxLayout,
    QLabel,
    QMainWindow,
    QScrollArea,
    QSplitter,
    QTabWidget,
    QVBoxLayout,
    QWidget,
)

from obc_controller.__version__ import __version__
from obc_controller.can_protocol import ChargerControl
from obc_controller.can_worker import BaudrateSwitchWorker, CANWorker
from obc_controller.simulator import Simulator
from obc_controller.ui.connection_panel import ConnectionPanel
from obc_controller.ui.control_panel import ControlPanel
from obc_controller.ui.graph_panel import GraphPanel
from obc_controller.ui.log_panel import LogPanel
from obc_controller.ui.telemetry_panel import TelemetryPanel
from obc_controller.ui.theme import COMPANY, ADDRESS, MADE_BY, CYAN, TEXT_DIM

_ASSETS = Path(__file__).parent / "assets"


class MainWindow(QMainWindow):
    def __init__(self) -> None:
        super().__init__()
        self.setWindowTitle(f"OBC Charger Controller \u2014 {COMPANY}")
        self.resize(1280, 860)

        self._worker: CANWorker | None = None
        self._simulator: Simulator | None = None
        self._baud_worker: BaudrateSwitchWorker | None = None
        self._sim_mode = False
        self._prev_control = ChargerControl.STOP_OUTPUTTING

        # --- central widget ---
        central = QWidget()
        central.setObjectName("central")
        self.setCentralWidget(central)
        root = QVBoxLayout(central)
        root.setContentsMargins(0, 0, 0, 0)
        root.setSpacing(0)

        # ── header ──────────────────────────────────────────────
        header = QFrame()
        header.setObjectName("header")
        header.setFixedHeight(64)
        h_lay = QHBoxLayout(header)
        h_lay.setContentsMargins(16, 4, 16, 4)

        logo_path = _ASSETS / "rtc_logo.png"
        if logo_path.exists():
            logo_lbl = QLabel()
            pm = QPixmap(str(logo_path)).scaled(
                48, 48, Qt.AspectRatioMode.KeepAspectRatio,
                Qt.TransformationMode.SmoothTransformation,
            )
            logo_lbl.setPixmap(pm)
            h_lay.addWidget(logo_lbl)

        title_col = QVBoxLayout()
        title_col.setSpacing(0)
        title_lbl = QLabel("OBC CHARGER CONTROLLER")
        title_lbl.setObjectName("header_title")
        subtitle_lbl = QLabel(f"{COMPANY}  \u00b7  {ADDRESS}")
        subtitle_lbl.setObjectName("header_subtitle")
        title_col.addWidget(title_lbl)
        title_col.addWidget(subtitle_lbl)
        h_lay.addLayout(title_col, stretch=1)

        about_btn_lbl = QLabel(
            f"<a style='color:{CYAN};' href='#'>About</a>"
        )
        about_btn_lbl.setCursor(Qt.CursorShape.PointingHandCursor)
        about_btn_lbl.linkActivated.connect(self._show_about)
        h_lay.addWidget(about_btn_lbl)

        root.addWidget(header)

        # ── 3-column body ───────────────────────────────────────
        body_splitter = QSplitter(Qt.Orientation.Horizontal)

        # Left sidebar — Connection
        left_widget = QWidget()
        left_widget.setObjectName("left_sidebar")
        left_widget.setMinimumWidth(240)
        left_widget.setMaximumWidth(340)
        left_inner = QVBoxLayout(left_widget)
        left_inner.setContentsMargins(6, 6, 6, 6)
        self._conn_panel = ConnectionPanel()
        left_inner.addWidget(self._conn_panel)
        left_inner.addStretch()

        left_scroll = QScrollArea()
        left_scroll.setWidgetResizable(True)
        left_scroll.setWidget(left_widget)
        left_scroll.setHorizontalScrollBarPolicy(
            Qt.ScrollBarPolicy.ScrollBarAlwaysOff
        )
        body_splitter.addWidget(left_scroll)

        # Center — Tabs (Graphs / Log) + Telemetry
        center = QWidget()
        center_lay = QVBoxLayout(center)
        center_lay.setContentsMargins(4, 4, 4, 4)

        tabs = QTabWidget()
        self._graph_panel = GraphPanel()
        self._log_panel = LogPanel()
        tabs.addTab(self._graph_panel, "Graphs")
        tabs.addTab(self._log_panel, "Log")
        center_lay.addWidget(tabs, stretch=3)

        self._tele_panel = TelemetryPanel()
        center_lay.addWidget(self._tele_panel, stretch=0)

        body_splitter.addWidget(center)

        # Right sidebar — Control
        right_widget = QWidget()
        right_widget.setObjectName("right_sidebar")
        right_widget.setMinimumWidth(260)
        right_widget.setMaximumWidth(380)
        right_inner = QVBoxLayout(right_widget)
        right_inner.setContentsMargins(6, 6, 6, 6)
        self._ctrl_panel = ControlPanel()
        right_inner.addWidget(self._ctrl_panel)
        right_inner.addStretch()

        right_scroll = QScrollArea()
        right_scroll.setWidgetResizable(True)
        right_scroll.setWidget(right_widget)
        right_scroll.setHorizontalScrollBarPolicy(
            Qt.ScrollBarPolicy.ScrollBarAlwaysOff
        )
        body_splitter.addWidget(right_scroll)

        # Splitter proportions: left 20%, center 55%, right 25%
        body_splitter.setStretchFactor(0, 2)
        body_splitter.setStretchFactor(1, 5)
        body_splitter.setStretchFactor(2, 2)
        root.addWidget(body_splitter, stretch=1)

        # ── footer ──────────────────────────────────────────────
        footer = QFrame()
        footer.setObjectName("footer")
        footer.setFixedHeight(28)
        f_lay = QHBoxLayout(footer)
        f_lay.setContentsMargins(12, 0, 12, 0)
        f_left = QLabel(f"{MADE_BY}  \u00b7  {COMPANY}, {ADDRESS}")
        f_left.setObjectName("footer_text")
        f_right = QLabel(f"v{__version__}")
        f_right.setObjectName("footer_text")
        f_lay.addWidget(f_left)
        f_lay.addStretch()
        f_lay.addWidget(f_right)
        root.addWidget(footer)

        # ── wire signals ────────────────────────────────────────
        self._conn_panel.connect_requested.connect(self._on_connect)
        self._conn_panel.disconnect_requested.connect(self._on_disconnect)
        self._conn_panel.baudrate_switch_requested.connect(
            self._on_baudrate_switch
        )
        self._ctrl_panel.control_changed.connect(self._on_control_changed)
        self._ctrl_panel.instant_360v_requested.connect(
            self._on_instant_360v
        )
        self._ctrl_panel.profile_loaded.connect(self._on_profile_loaded)
        self._ctrl_panel.log_message.connect(self._log_panel.append)

    # ---- About dialog ----------------------------------------------------

    def _show_about(self) -> None:
        dlg = QDialog(self)
        dlg.setWindowTitle("About OBC Charger Controller")
        dlg.setFixedSize(420, 320)
        lay = QVBoxLayout(dlg)
        lay.setAlignment(Qt.AlignmentFlag.AlignCenter)
        lay.setSpacing(12)

        logo_path = _ASSETS / "rtc_logo.png"
        if logo_path.exists():
            logo_lbl = QLabel()
            pm = QPixmap(str(logo_path)).scaled(
                96, 96, Qt.AspectRatioMode.KeepAspectRatio,
                Qt.TransformationMode.SmoothTransformation,
            )
            logo_lbl.setPixmap(pm)
            logo_lbl.setAlignment(Qt.AlignmentFlag.AlignCenter)
            lay.addWidget(logo_lbl)

        title = QLabel("OBC Charger Controller")
        title.setStyleSheet(
            f"font-size: 18px; font-weight: bold; color: {CYAN};"
        )
        title.setAlignment(Qt.AlignmentFlag.AlignCenter)
        lay.addWidget(title)

        ver = QLabel(f"Version {__version__}")
        ver.setAlignment(Qt.AlignmentFlag.AlignCenter)
        lay.addWidget(ver)

        company = QLabel(f"{COMPANY}\n{ADDRESS}")
        company.setAlignment(Qt.AlignmentFlag.AlignCenter)
        company.setStyleSheet(f"color: {TEXT_DIM};")
        lay.addWidget(company)

        made = QLabel(MADE_BY)
        made.setAlignment(Qt.AlignmentFlag.AlignCenter)
        made.setStyleSheet(f"color: {CYAN}; font-weight: bold;")
        lay.addWidget(made)

        lay.addStretch()
        dlg.exec()

    # ---- connection management -------------------------------------------

    @Slot(str, str, int, bool)
    def _on_connect(
        self, interface: str, channel: str, bitrate: int, simulate: bool
    ) -> None:
        self._sim_mode = simulate

        if simulate:
            self._log_panel.append("Starting simulation mode \u2026")
            self._simulator = Simulator()
            self._simulator.message2_received.connect(self._on_message2)
            self._simulator.log_message.connect(self._log_panel.append)
            self._simulator.start()
            self._conn_panel.set_connected(True)
            self._ctrl_panel.setEnabled(True)
            # Set initial setpoints in telemetry for sim mode
            self._tele_panel.update_setpoints(
                self._ctrl_panel.get_voltage(),
                self._ctrl_panel.get_current(),
            )
            return

        # Real CAN connection
        self._worker = CANWorker()
        self._worker.set_connection_params(interface, channel, bitrate)

        # Apply initial setpoints + ramp config
        self._worker.set_setpoints(
            self._ctrl_panel.get_voltage(),
            self._ctrl_panel.get_current(),
        )
        self._worker.set_control(self._ctrl_panel.get_control())
        ramp_v, ramp_a = self._ctrl_panel.get_ramp_rates()
        self._worker.set_ramp_config(
            self._ctrl_panel.get_ramp_enabled(), ramp_v, ramp_a
        )
        self._worker.enable_tx(True)

        # Wire worker signals
        self._worker.connected.connect(self._on_worker_connected)
        self._worker.disconnected.connect(self._on_worker_disconnected)
        self._worker.error.connect(self._on_worker_error)
        self._worker.log_message.connect(self._log_panel.append)
        self._worker.message2_received.connect(self._on_message2)
        self._worker.timeout_alarm.connect(self._on_timeout_alarm)
        self._worker.tx_message.connect(self._on_tx_message)
        self._worker.ramp_state.connect(self._on_ramp_state)
        self._worker.health_stats.connect(self._on_health_stats)
        self._worker.status_bit_changed.connect(self._on_status_bit_changed)

        self._worker.start()

    @Slot()
    def _on_disconnect(self) -> None:
        if self._sim_mode and self._simulator is not None:
            self._simulator.request_stop()
            self._simulator.wait(3000)
            self._simulator = None
            self._conn_panel.set_connected(False)
            self._ctrl_panel.setEnabled(False)
            self._tele_panel.clear()
            self._log_panel.append("Simulation stopped.")
            return

        if self._worker is not None:
            self._log_panel.append("Disconnecting \u2026")
            self._worker.request_stop()
            self._worker.wait(10000)
            self._worker = None

    # ---- worker signal handlers ------------------------------------------

    @Slot()
    def _on_worker_connected(self) -> None:
        self._conn_panel.set_connected(True)
        self._ctrl_panel.setEnabled(True)

    @Slot()
    def _on_worker_disconnected(self) -> None:
        self._conn_panel.set_connected(False)
        self._ctrl_panel.setEnabled(False)
        self._tele_panel.clear()
        self._ctrl_panel.update_ramp_display(False, 0, 0)

    @Slot(str)
    def _on_worker_error(self, msg: str) -> None:
        self._log_panel.append(f"ERROR: {msg}")

    @Slot(object)
    def _on_message2(self, msg) -> None:
        self._tele_panel.update_telemetry(msg)
        self._graph_panel.add_point(msg)

    @Slot()
    def _on_timeout_alarm(self) -> None:
        self._tele_panel.set_alarm("ALARM: No Message2 > 5 s!")

    @Slot(object)
    def _on_tx_message(self, msg) -> None:
        self._log_panel.append(
            f"TX Message1: V={msg.voltage_setpoint:.1f}V "
            f"I={msg.current_setpoint:.1f}A "
            f"ctrl={msg.control.name}"
        )
        # Feed SET values to telemetry (ramped or target, whichever is sent)
        self._tele_panel.update_setpoints(
            msg.voltage_setpoint, msg.current_setpoint
        )

    @Slot(bool, float, float)
    def _on_ramp_state(
        self, active: bool, ramped_v: float, ramped_a: float
    ) -> None:
        self._ctrl_panel.update_ramp_display(active, ramped_v, ramped_a)

    @Slot(float, float, float)
    def _on_health_stats(
        self, tx_rate: float, rx_rate: float, last_rx_age: float
    ) -> None:
        self._conn_panel.update_health(tx_rate, rx_rate, last_rx_age)

    @Slot(int, str, bool)
    def _on_status_bit_changed(
        self, bit: int, name: str, is_fault: bool
    ) -> None:
        state = "ON" if is_fault else "OFF"
        severity = "error" if is_fault else "info"
        self._graph_panel.add_event_marker(
            f"FAULT {name} {state}", severity
        )
        self._log_panel.append(f"Status bit {bit} ({name}): {state}")

    # ---- control panel -> worker -----------------------------------------

    @Slot(float, float, int, bool, float, float)
    def _on_control_changed(
        self,
        voltage: float,
        current: float,
        control: int,
        ramp_enabled: bool,
        ramp_v: float,
        ramp_a: float,
    ) -> None:
        ctrl = ChargerControl(control)

        # Detect mode change → add graph marker
        if ctrl != self._prev_control:
            label = {
                ChargerControl.STOP_OUTPUTTING: "STOP",
                ChargerControl.START_CHARGING: "START CHARGING",
                ChargerControl.HEATING_DC_SUPPLY: "HEATING/DC",
            }.get(ctrl, "?")
            self._graph_panel.add_event_marker(f"Mode: {label}", "info")
            self._prev_control = ctrl

        # Update telemetry setpoints in sim mode (no tx_message signal)
        if self._sim_mode:
            self._tele_panel.update_setpoints(voltage, current)

        if self._worker is not None:
            self._worker.set_setpoints(voltage, current)
            self._worker.set_control(ctrl)
            self._worker.set_ramp_config(ramp_enabled, ramp_v, ramp_a)

    @Slot(object)
    def _on_profile_loaded(self, profile) -> None:
        """Reset ramp when a profile is loaded."""
        if self._worker is not None:
            self._worker.reset_ramp()

    # ---- instant 360V / 9A -----------------------------------------------

    @Slot()
    def _on_instant_360v(self) -> None:
        """Apply 360V/9A instant preset to the worker."""
        self._graph_panel.add_event_marker(
            "\u26a1 360V/9A INSTANT", "warning"
        )
        if self._worker is not None:
            self._worker.set_setpoints(360.0, 9.0)
            self._worker.set_control(ChargerControl.HEATING_DC_SUPPLY)
            self._worker.set_ramp_config(False, 5.0, 0.5)
            self._worker.reset_ramp()

    # ---- baudrate switch sequence ----------------------------------------

    @Slot()
    def _on_baudrate_switch(self) -> None:
        if self._worker is None or self._worker._bus is None:
            self._log_panel.append(
                "Cannot switch baudrate: CAN not connected."
            )
            return
        if self._baud_worker is not None and self._baud_worker.isRunning():
            self._log_panel.append(
                "Baudrate switch already in progress."
            )
            return

        self._conn_panel.set_baud_switch_busy(True)
        self._ctrl_panel.setEnabled(False)
        self._graph_panel.add_event_marker("Baud switch START", "info")

        self._baud_worker = BaudrateSwitchWorker(self._worker._bus)
        self._baud_worker.progress.connect(self._on_baud_progress)
        self._baud_worker.finished_ok.connect(self._on_baud_done)
        self._baud_worker.error.connect(self._on_baud_error)
        self._baud_worker.log_message.connect(self._log_panel.append)
        self._baud_worker.start()

    @Slot(int, int)
    def _on_baud_progress(self, step: int, total: int) -> None:
        self._conn_panel.set_baud_switch_progress(step, total)

    @Slot()
    def _on_baud_done(self) -> None:
        self._conn_panel.set_baud_switch_done()
        self._ctrl_panel.setEnabled(True)
        self._baud_worker = None
        self._graph_panel.add_event_marker("Baud switch DONE", "info")

    @Slot(str)
    def _on_baud_error(self, msg: str) -> None:
        self._log_panel.append(f"ERROR: {msg}")
        self._conn_panel.set_baud_switch_busy(False)
        self._ctrl_panel.setEnabled(True)
        self._baud_worker = None

    # ---- window close ----------------------------------------------------

    def closeEvent(self, event) -> None:
        if self._baud_worker is not None and self._baud_worker.isRunning():
            self._baud_worker.request_stop()
            self._baud_worker.wait(5000)
            self._baud_worker = None
        self._on_disconnect()
        super().closeEvent(event)

File: cpp_app/ui/main_window.cpp
#include "ui/main_window.h"
#include "ui/theme.h"

#include <QCloseEvent>
#include <QDialog>
#include <QFrame>
#include <QHBoxLayout>
#include <QLabel>
#include <QScrollArea>
#include <QSplitter>
#include <QTabWidget>
#include <QVBoxLayout>

MainWindow::MainWindow(QWidget* parent)
    : QMainWindow(parent)
{
    setWindowTitle(QString("OBC Charger Controller \u2014 %1").arg(Theme::COMPANY));
    resize(1280, 860);

    // Central widget
    auto* central = new QWidget;
    central->setObjectName("central");
    setCentralWidget(central);
    auto* root = new QVBoxLayout(central);
    root->setContentsMargins(0, 0, 0, 0);
    root->setSpacing(0);

    // Header
    auto* header = new QFrame;
    header->setObjectName("header");
    header->setFixedHeight(64);
    auto* hLay = new QHBoxLayout(header);
    hLay->setContentsMargins(16, 4, 16, 4);

    auto* titleCol = new QVBoxLayout;
    titleCol->setSpacing(0);
    auto* titleLbl = new QLabel("OBC CHARGER CONTROLLER");
    titleLbl->setObjectName("header_title");
    auto* subtitleLbl = new QLabel(QString("%1  \u00b7  %2").arg(Theme::COMPANY, Theme::ADDRESS));
    subtitleLbl->setObjectName("header_subtitle");
    titleCol->addWidget(titleLbl);
    titleCol->addWidget(subtitleLbl);
    hLay->addLayout(titleCol, 1);

    auto* aboutLbl = new QLabel(QString("<a style='color:%1;' href='#'>About</a>").arg(Theme::CYAN));
    aboutLbl->setCursor(Qt::PointingHandCursor);
    connect(aboutLbl, &QLabel::linkActivated, this, &MainWindow::showAbout);
    hLay->addWidget(aboutLbl);
    root->addWidget(header);

    // 3-column body
    auto* bodySplitter = new QSplitter(Qt::Horizontal);

    // Left sidebar - Connection
    auto* leftWidget = new QWidget;
    leftWidget->setObjectName("left_sidebar");
    leftWidget->setMinimumWidth(240);
    leftWidget->setMaximumWidth(340);
    auto* leftInner = new QVBoxLayout(leftWidget);
    leftInner->setContentsMargins(6, 6, 6, 6);
    m_connPanel = new ConnectionPanel;
    leftInner->addWidget(m_connPanel);
    leftInner->addStretch();

    auto* leftScroll = new QScrollArea;
    leftScroll->setWidgetResizable(true);
    leftScroll->setWidget(leftWidget);
    leftScroll->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    bodySplitter->addWidget(leftScroll);

    // Center - Tabs + Telemetry
    auto* centerWidget = new QWidget;
    auto* centerLay = new QVBoxLayout(centerWidget);
    centerLay->setContentsMargins(4, 4, 4, 4);

    auto* tabs = new QTabWidget;
    m_graphPanel = new GraphPanel;
    m_logPanel = new LogPanel;
    tabs->addTab(m_graphPanel, "Graphs");
    tabs->addTab(m_logPanel, "Log");
    centerLay->addWidget(tabs, 3);

    m_telePanel = new TelemetryPanel;
    centerLay->addWidget(m_telePanel, 0);
    bodySplitter->addWidget(centerWidget);

    // Right sidebar - Control
    auto* rightWidget = new QWidget;
    rightWidget->setObjectName("right_sidebar");
    rightWidget->setMinimumWidth(260);
    rightWidget->setMaximumWidth(380);
    auto* rightInner = new QVBoxLayout(rightWidget);
    rightInner->setContentsMargins(6, 6, 6, 6);
    m_ctrlPanel = new ControlPanel;
    rightInner->addWidget(m_ctrlPanel);
    rightInner->addStretch();

    auto* rightScroll = new QScrollArea;
    rightScroll->setWidgetResizable(true);
    rightScroll->setWidget(rightWidget);
    rightScroll->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    bodySplitter->addWidget(rightScroll);

    bodySplitter->setStretchFactor(0, 2);
    bodySplitter->setStretchFactor(1, 5);
    bodySplitter->setStretchFactor(2, 2);
    root->addWidget(bodySplitter, 1);

    // Footer
    auto* footer = new QFrame;
    footer->setObjectName("footer");
    footer->setFixedHeight(28);
    auto* fLay = new QHBoxLayout(footer);
    fLay->setContentsMargins(12, 0, 12, 0);
    auto* fLeft = new QLabel(QString("%1  \u00b7  %2, %3")
                                 .arg(Theme::MADE_BY, Theme::COMPANY, Theme::ADDRESS));
    fLeft->setObjectName("footer_text");
    auto* fRight = new QLabel(QString("v%1").arg(Theme::VERSION));
    fRight->setObjectName("footer_text");
    fLay->addWidget(fLeft);
    fLay->addStretch();
    fLay->addWidget(fRight);
    root->addWidget(footer);

    // Wire signals
    connect(m_connPanel, &ConnectionPanel::connectRequested, this, &MainWindow::onConnect);
    connect(m_connPanel, &ConnectionPanel::disconnectRequested, this, &MainWindow::onDisconnect);
    connect(m_connPanel, &ConnectionPanel::baudrateSwitchRequested, this, &MainWindow::onBaudrateSwitch);
    connect(m_ctrlPanel, &ControlPanel::controlChanged, this, &MainWindow::onControlChanged);
    connect(m_ctrlPanel, &ControlPanel::instant360vRequested, this, &MainWindow::onInstant360v);
    connect(m_ctrlPanel, &ControlPanel::profileLoaded, this, &MainWindow::onProfileLoaded);
    connect(m_ctrlPanel, &ControlPanel::logMessage, m_logPanel, &LogPanel::append);
}

// --- About ---

void MainWindow::showAbout() {
    auto* dlg = new QDialog(this);
    dlg->setWindowTitle("About OBC Charger Controller");
    dlg->setFixedSize(420, 280);
    auto* lay = new QVBoxLayout(dlg);
    lay->setAlignment(Qt::AlignCenter);
    lay->setSpacing(12);

    auto* title = new QLabel("OBC Charger Controller");
    title->setStyleSheet(QString("font-size: 18px; font-weight: bold; color: %1;").arg(Theme::CYAN));
    title->setAlignment(Qt::AlignCenter);
    lay->addWidget(title);

    auto* ver = new QLabel(QString("Version %1 (C++)").arg(Theme::VERSION));
    ver->setAlignment(Qt::AlignCenter);
    lay->addWidget(ver);

    auto* company = new QLabel(QString("%1\n%2").arg(Theme::COMPANY, Theme::ADDRESS));
    company->setAlignment(Qt::AlignCenter);
    company->setStyleSheet(QString("color: %1;").arg(Theme::TEXT_DIM));
    lay->addWidget(company);

    auto* made = new QLabel(Theme::MADE_BY);
    made->setAlignment(Qt::AlignCenter);
    made->setStyleSheet(QString("color: %1; font-weight: bold;").arg(Theme::CYAN));
    lay->addWidget(made);

    lay->addStretch();
    dlg->exec();
    delete dlg;
}

// --- Connection ---

void MainWindow::onConnect(const QString& interface, const QString& channel,
                           int bitrate, bool simulate)
{
    m_simMode = simulate;

    if (simulate) {
        m_logPanel->append(QString::fromUtf8("Starting simulation mode \u2026"));
        m_simulator = new Simulator;
        connect(m_simulator, &Simulator::message2Received, this, &MainWindow::onMessage2);
        connect(m_simulator, &Simulator::logMessage, m_logPanel, &LogPanel::append);
        m_simulator->start();
        m_connPanel->setConnected(true);
        m_ctrlPanel->setEnabled(true);
        m_telePanel->updateSetpoints(m_ctrlPanel->getVoltage(), m_ctrlPanel->getCurrent());
        return;
    }

    m_worker = new CANWorker;
    m_worker->setConnectionParams(interface, channel, bitrate);
    m_worker->setSetpoints(m_ctrlPanel->getVoltage(), m_ctrlPanel->getCurrent());
    m_worker->setControl(m_ctrlPanel->getControl());
    auto [rampV, rampA] = m_ctrlPanel->getRampRates();
    m_worker->setRampConfig(m_ctrlPanel->getRampEnabled(), rampV, rampA);
    m_worker->enableTx(true);

    connect(m_worker, &CANWorker::connected, this, &MainWindow::onWorkerConnected);
    connect(m_worker, &CANWorker::disconnected, this, &MainWindow::onWorkerDisconnected);
    connect(m_worker, &CANWorker::error, this, &MainWindow::onWorkerError);
    connect(m_worker, &CANWorker::logMessage, m_logPanel, &LogPanel::append);
    connect(m_worker, &CANWorker::message2Received, this, &MainWindow::onMessage2);
    connect(m_worker, &CANWorker::timeoutAlarm, this, &MainWindow::onTimeoutAlarm);
    connect(m_worker, &CANWorker::txMessage, this, &MainWindow::onTxMessage);
    connect(m_worker, &CANWorker::rampState, this, &MainWindow::onRampState);
    connect(m_worker, &CANWorker::healthStats, this, &MainWindow::onHealthStats);
    connect(m_worker, &CANWorker::statusBitChanged, this, &MainWindow::onStatusBitChanged);

    m_worker->start();
}

void MainWindow::onDisconnect() {
    if (m_simMode && m_simulator) {
        m_simulator->requestStop();
        m_simulator->wait(3000);
        delete m_simulator;
        m_simulator = nullptr;
        m_connPanel->setConnected(false);
        m_ctrlPanel->setEnabled(false);
        m_telePanel->clear();
        m_logPanel->append("Simulation stopped.");
        return;
    }

    if (m_worker) {
        m_logPanel->append(QString::fromUtf8("Disconnecting \u2026"));
        m_worker->requestStop();
        m_worker->wait(10000);
        delete m_worker;
        m_worker = nullptr;
    }
}

// --- Worker signal handlers ---

void MainWindow::onWorkerConnected() {
    m_connPanel->setConnected(true);
    m_ctrlPanel->setEnabled(true);
}

void MainWindow::onWorkerDisconnected() {
    m_connPanel->setConnected(false);
    m_ctrlPanel->setEnabled(false);
    m_telePanel->clear();
    m_ctrlPanel->updateRampDisplay(false, 0, 0);
}

void MainWindow::onWorkerError(const QString& msg) {
    m_logPanel->append("ERROR: " + msg);
}

void MainWindow::onMessage2(Message2 msg) {
    m_telePanel->updateTelemetry(msg);
    m_graphPanel->addPoint(msg);
}

void MainWindow::onTimeoutAlarm() {
    m_telePanel->setAlarm("ALARM: No Message2 > 5 s!");
}

void MainWindow::onTxMessage(Message1 msg) {
    m_logPanel->append(QString("TX Message1: V=%1V I=%2A ctrl=%3")
                           .arg(msg.voltage_setpoint, 0, 'f', 1)
                           .arg(msg.current_setpoint, 0, 'f', 1)
                           .arg(QString::fromStdString(chargerControlName(msg.control))));
    m_telePanel->updateSetpoints(msg.voltage_setpoint, msg.current_setpoint);
}

void MainWindow::onRampState(bool active, double rampedV, double rampedA) {
    m_ctrlPanel->updateRampDisplay(active, rampedV, rampedA);
}

void MainWindow::onHealthStats(double txRate, double rxRate, double lastRxAge) {
    m_connPanel->updateHealth(txRate, rxRate, lastRxAge);
}

void MainWindow::onStatusBitChanged(int bit, const QString& name, bool isFault) {
    QString state = isFault ? "ON" : "OFF";
    QString severity = isFault ? "error" : "info";
    m_graphPanel->addEventMarker(QString("FAULT %1 %2").arg(name, state), severity);
    m_logPanel->append(QString("Status bit %1 (%2): %3").arg(bit).arg(name, state));
}

// --- Control panel -> worker ---

void MainWindow::onControlChanged(double voltage, double current, int control,
                                  bool rampEnabled, double rampV, double rampA)
{
    auto ctrl = static_cast<ChargerControl>(control);

    if (ctrl != m_prevControl) {
        QMap<ChargerControl, QString> labelMap = {
            {ChargerControl::STOP_OUTPUTTING,   "STOP"},
            {ChargerControl::START_CHARGING,    "START CHARGING"},
            {ChargerControl::HEATING_DC_SUPPLY, "HEATING/DC"},
        };
        QString label = labelMap.value(ctrl, "?");
        m_graphPanel->addEventMarker("Mode: " + label, "info");
        m_prevControl = ctrl;
    }

    if (m_simMode)
        m_telePanel->updateSetpoints(voltage, current);

    if (m_worker) {
        m_worker->setSetpoints(voltage, current);
        m_worker->setControl(ctrl);
        m_worker->setRampConfig(rampEnabled, rampV, rampA);
    }
}

void MainWindow::onProfileLoaded() {
    if (m_worker)
        m_worker->resetRamp();
}

// --- Instant 360V ---

void MainWindow::onInstant360v() {
    m_graphPanel->addEventMarker(QString::fromUtf8("\u26a1 360V/9A INSTANT"), "warning");
    if (m_worker) {
        m_worker->setSetpoints(360.0, 9.0);
        m_worker->setControl(ChargerControl::HEATING_DC_SUPPLY);
        m_worker->setRampConfig(false, 5.0, 0.5);
        m_worker->resetRamp();
    }
}

// --- Baudrate switch ---

void MainWindow::onBaudrateSwitch() {
    if (!m_worker || !m_worker->isBusConnected()) {
        m_logPanel->append("Cannot switch baudrate: CAN not connected.");
        return;
    }
    if (m_baudWorker && m_baudWorker->isRunning()) {
        m_logPanel->append("Baudrate switch already in progress.");
        return;
    }

    m_connPanel->setBaudSwitchBusy(true);
    m_ctrlPanel->setEnabled(false);
    m_graphPanel->addEventMarker("Baud switch START", "info");

    m_baudWorker = new BaudrateSwitchWorker(m_worker->getSocketFd());
    connect(m_baudWorker, &BaudrateSwitchWorker::progress, this, &MainWindow::onBaudProgress);
    connect(m_baudWorker, &BaudrateSwitchWorker::finishedOk, this, &MainWindow::onBaudDone);
    connect(m_baudWorker, &BaudrateSwitchWorker::error, this, &MainWindow::onBaudError);
    connect(m_baudWorker, &BaudrateSwitchWorker::logMessage, m_logPanel, &LogPanel::append);
    m_baudWorker->start();
}

void MainWindow::onBaudProgress(int step, int total) {
    m_connPanel->setBaudSwitchProgress(step, total);
}

void MainWindow::onBaudDone() {
    m_connPanel->setBaudSwitchDone();
    m_ctrlPanel->setEnabled(true);
    delete m_baudWorker;
    m_baudWorker = nullptr;
    m_graphPanel->addEventMarker("Baud switch DONE", "info");
}

void MainWindow::onBaudError(const QString& msg) {
    m_logPanel->append("ERROR: " + msg);
    m_connPanel->setBaudSwitchBusy(false);
    m_ctrlPanel->setEnabled(true);
    delete m_baudWorker;
    m_baudWorker = nullptr;
}

// --- Close ---

void MainWindow::closeEvent(QCloseEvent* event) {
    if (m_baudWorker && m_baudWorker->isRunning()) {
        m_baudWorker->requestStop();
        m_baudWorker->wait(5000);
        delete m_baudWorker;
        m_baudWorker = nullptr;
    }
    onDisconnect();
    QMainWindow::closeEvent(event);
}
